test/colony-network-recovery.js

```diff
@@ -0,0 +1,498 @@
+            loader: contractLoader,
+            minerAddress: accounts[5],
+            realProviderPort: REAL_PROVIDER_PORT,
+            useJsTree: true
+          });
+          await newMiningClient.initialise(colonyNetwork.address);
+
+          const tokenArgs = getTokenArgs();
+          const token = await Token.new(...tokenArgs);
+          const { logs } = await colonyNetwork.createColony(token.address);
+          const { colonyAddress } = logs[0].args;
+
+          await token.setOwner(colonyAddress);
+          const colony = await IColony.at(colonyAddress);
+
+          await colony.mintTokens(1000000000000000);
+          await colony.bootstrapColony([accounts[0]], [1000000000000000]);
+
+          let addr = await colonyNetwork.getReputationMiningCycle.call(false);
+          let repCycle = await IReputationMiningCycle.at(addr);
+
+          await miningClient.addLogContentsToReputationTree();
+          await forwardTime(3600, this);
+          await miningClient.submitRootHash();
+
+          addr = await colonyNetwork.getReputationMiningCycle.call(true);
+          repCycle = await IReputationMiningCycle.at(addr);
+          await repCycle.confirmNewHash(0);
+
+          await miningClient.addLogContentsToReputationTree();
+          await forwardTime(3600, this);
+          await miningClient.submitRootHash();
+
+          addr = await colonyNetwork.getReputationMiningCycle.call(true);
+          repCycle = await IReputationMiningCycle.at(addr);
+
+          const invalidEntry = await repCycle.getReputationUpdateLogEntry(5);
+          invalidEntry.amount = 0;
+
+          await repCycle.confirmNewHash(0);
+
+          const domain = await colony.getDomain(1);
+          const rootSkill = domain[0];
+          const reputationKey = makeReputationKey(colony.address, rootSkill.toNumber(), accounts[0]);
+          const originalValue = miningClient.reputations[reputationKey].slice(2, 66);
+          assert.equal(parseInt(originalValue, 16), 1000000000000000);
+
+          await colonyNetwork.enterRecoveryMode();
+
+          await colonyNetwork.setReplacementReputationUpdateLogEntry(
+            addr,
+            5,
+            invalidEntry.user,
+            invalidEntry.amount,
+            invalidEntry.skillId,
+            invalidEntry.colony,
+            invalidEntry.nUpdates,
+            invalidEntry.nPreviousUpdates
+          );
+          await miningClient.loadState(startingHash);
+          // This sync call will log an error - this is because we've changed a log entry, but the root hash
+          // on-chain which .sync() does a sanity check against hasn't been updated.
+          await miningClient.sync(startingBlockNumber, true);
+          console.log("The WARNING and ERROR immediately preceeding can be ignored (they are expected as part of the test)");
+          const rootHash = await miningClient.getRootHash();
+          const nNodes = await miningClient.nReputations;
```

Comment:
We get the "right" root hash and nNodes and set them on the network to correct the "wrong" ones from the previous update.